Implement OTEP 0178: Mapping external data to AnyValue (#2385)
open-telemetry/opentelemetry-specification: specification/common/attribute-type-mapping.md

# Mapping Arbitrary Data to OTLP AnyValue

**Status**: [Experimental](../document-status.md)

<details>
<summary>Table of Contents</summary>

<!-- toc -->

- [Converting to AnyValue](#converting-to-anyvalue)
  * [Primitive Values](#primitive-values)
    + [Integer Values](#integer-values)
    + [Enumerations](#enumerations)
    + [Floating Point Values](#floating-point-values)
    + [String Values](#string-values)
    + [Bytes Sequences](#bytes-sequences)
  * [Composite Values](#composite-values)
    + [Array Values](#array-values)
    + [Associative Arrays With Unique Keys](#associative-arrays-with-unique-keys)
    + [Associative Arrays With Non-Unique Keys](#associative-arrays-with-non-unique-keys)
    + [Sets](#sets)
  * [Other Values](#other-values)
  * [Empty Values](#empty-values)

<!-- tocstop -->

</details>

This document defines how to map (convert) arbitrary data (e.g. in-memory
objects) to OTLP's [AnyValue](https://github.com/open-telemetry/opentelemetry-proto/blob/cc4ed55c082cb75e084d40b4ddf3805eda099f97/opentelemetry/proto/common/v1/common.proto#L27).

The mapping is needed when OpenTelemetry needs to convert a value produced outside

> OpenTelemetry into a value that can be exported using OTLP exporter, or otherwise be

Reviewer: ```suggestion
OpenTelemetry into a value that can be exported using an OTLP exporter, or otherwise be
```

Author: Fixed.

converted to be used inside OpenTelemetry boundaries. Example use cases are the following:

> - In [Logging Library SDK](../logs/logging-library-sdk.md) to convert values received

Reviewer: ```suggestion
- In [Logging Library SDK](../logs/logging-library-sdk.md)s, to convert values received
```

Author: Fixed.

  from logging libraries into OpenTelemetry representation.

> - In the Collector to convert values received from various data sources into

Reviewer: ```suggestion
- In the Collector, to convert values received from various data sources into the
```

Author: Fixed.

  [pdata](https://github.com/open-telemetry/opentelemetry-collector/blob/4998703dadd19fa91a88eabf7ccc68d728bee3fd/model/pdata/common.go#L84)
  internal representation.

## Converting to AnyValue

[AnyValue](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L27)
is capable of representing primitive and structured data of certain types.

> Implementations that have a source data in any form, such as in-memory objects

Reviewer: ```suggestion
Implementations that have source data in any form, such as in-memory objects
```

Author: Fixed.

or data coming from other formats that needs to be converted to AnyValue SHOULD
follow the rules described below.

### Primitive Values

#### Integer Values

Integer values which are within the range of 64 bit signed numbers
[-2^63..2^63-1] SHOULD be converted to AnyValue's
[int_value](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L33)
field.

Integer values which are outside the range of 64 bit signed numbers SHOULD be
converted to AnyValue's
[string_value](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L31)
field using decimal representation.

#### Enumerations

Values, which belong to a limited enumerated set (e.g. a Java

> [enum](https://docs.oracle.com/javase/tutorial/java/javaOO/enum.html)) SHOULD be

Reviewer: ```suggestion
[enum](https://docs.oracle.com/javase/tutorial/java/javaOO/enum.html)), SHOULD be
```

Author: Fixed.

converted to AnyValue's
[string_value](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L31)
field with the value of the string set to the symbolic name of the enumeration.

> If the symbolic name of the enumeration is not possible to obtain the

Reviewer: ```suggestion
If it is not possible to obtain the symbolic name of the enumeration, the
```

> implementation SHOULD map enumeration's value to AnyValue's

Reviewer: ```suggestion
implementation SHOULD map enumeration values to AnyValue's
```

Author: Fixed.

[int_value](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L33)

> field set equal to enum's ordinal number when such ordinal number is naturally

Reviewer: ```suggestion
field set to the enum's ordinal value, when such an ordinal number is naturally
```

Author: Fixed.

obtainable.

> If the ordinal value is also not possible to obtain the enumeration SHOULD be

Reviewer: ```suggestion
If it is also not possible to obtain the ordinal value, the enumeration SHOULD be
```

Author: Fixed.

[double_value](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L34)
field.

Floating point values which are outside the range or precision of IEEE 754
64-bit floating point numbers (e.g. IEEE 128-bit floating bit values) SHOULD be
converted to AnyValue's
[string_value](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L31)
field using decimal floating point representation.

#### String Values

String values which are valid Unicode sequences SHOULD be converted to
AnyValue's
[string_value](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L31)
field.

String values which are not valid Unicode sequences SHOULD be converted to
AnyValue's
[bytes_value](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L37)
with the bytes representing the string in the original order and format of the
source string.

> #### Bytes Sequences

Author: Fixed.

Byte sequences (e.g. Go's `[]byte` slice or raw byte content of a file) SHOULD
be converted to AnyValue's
[bytes_value](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L37)
field.

### Composite Values

#### Array Values

Values that represent ordered sequences of other values (such as
[arrays](https://docs.oracle.com/javase/specs/jls/se7/html/jls-10.html),
[vectors](https://en.cppreference.com/w/cpp/container/vector), ordered
[lists](https://docs.python.org/3/tutorial/datastructures.html#more-on-lists),
[slices](https://golang.org/ref/spec#Slice_types)) SHOULD be converted to
AnyValue's
[array_value](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L35)

> field. String Values and Byte Sequences are an exception from this rule (see

Reviewer: ```suggestion
field. String values and byte sequences are an exception from this rule (see
```

Author: Fixed.

above).

> 

Reviewer: ```suggestion
OTLP's `ArrayValue` consists of `AnyValue` types. Therefore, it is possible to have arrays of mixed types in OTLP. When converting an array to OTLP, the rules described in this document should also be applied to each element in the source array. 

Arrays can be arbitrarily nested.

```
I am on the fence about this one, since I think it can lead to very large and very nested structures in the Arrays, but I don't really see a way of restricting this.

Author: I added a sentence to explain that the rules should be applied recursively. That's definitely the intent.

#### Associative Arrays With Unique Keys

Values that represent associative arrays with unique keys (also often known
as maps, dictionaries or key-value stores) SHOULD be converted to AnyValue's
[kvlist_value](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L36)
field.

> If the keys of the source array are not strings they MUST be converted to

Reviewer: ```suggestion
If the keys of the source array are not strings, they MUST be converted to
```

Author: Fixed.

field MUST be an array represented using AnyValue's
[array_value](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L35)
field, each element of the
[array_value](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L35)
representing one value of source array associated with the given key.

For example an associative array shown in the following table:

|Key|Value|
|---|---|
|"abc"|123|
|"def"|"foo"|
|"def"|"bar"|

SHOULD be converted to:

```
AnyValue{
    kvlist_value:KeyValueList{
        values:[
            KeyValue{
                key:"abc",
                value:AnyValue{array_value:ArrayValue{values[
                    AnyValue{int_value:123}
                ]}}
            },
            KeyValue{
                key:"def",
                value:AnyValue{array_value:ArrayValue{values[
                    AnyValue{string_value:"foo"},
                    AnyValue{string_value:"bar"}
                ]}}
            },
        ]
    }
}
```

#### Sets

> Unordered collections of non-duplicate values (such as

Reviewer: ```suggestion
Unordered collections of unique values (such as
```
To keep in line with the wording above

Author: Fixed.

obtainable.

If it is also not possible to obtain the ordinal value, the enumeration SHOULD be
converted to AnyValue's
[bytes_value](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L37)
field in any manner that the implementation deems reasonable.

#### Floating Point Values

Floating point values which are within the range and precision of IEEE 754
64-bit floating point numbers (including IEEE 32-bit floating point values)
SHOULD be converted to AnyValue's
[double_value](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L34)
field.

Floating point values which are outside the range or precision of IEEE 754
64-bit floating point numbers (e.g. IEEE 128-bit floating bit values) SHOULD be
converted to AnyValue's
[string_value](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L31)
field using decimal floating point representation.

#### String Values

> String values which are valid Unicode sequences SHOULD be converted to

Reviewer 2: > valid Unicode sequences

Should we be more strict and say "valid Unicode UTF-8 sequences"? Since we don't capture the encoding anywhere else, if the encoding is not known then "valid Unicode sequences" cannot be determined.
